From 16005de55b1bb53f963688163d6181f9e929a88d Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Fri, 2 Jun 2023 12:18:58 -0700
Subject: [PATCH] ath12k: Clear the flush flags after ce init  and add few more
  checks for checking flush flags

    Usually when work threads carry out the process of
    reconfiguring the chip which was asserted, a lot of other
    things can happen from upper layers, requesting some
    actions especially when the chip is down. Adding a check
    for chip recovery might help at times, but it need not
    always work because of this worker thread async.

    This change will be checking the chip recovery flag in the mac_op functions.
    If the flag ATH12K_FLAG_CRASH_FLUSH is set then return -ESTUTDOWN.

    This will ensure, when the  restart_work and reset_work threads
    take care of the chip recovery and reset work respectively, they set and
    clear the FLUSH bits at the right places, when the FW is fully down and when
    the FW has been initialized properly and the upper layers such as mac80211
    get failure messages when they send out WMI commands when the chip is
    still down.

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c |  4 ++--
 drivers/net/wireless/ath/ath12k/core.h |  1 +
 drivers/net/wireless/ath/ath12k/htc.c  |  2 ++
 drivers/net/wireless/ath/ath12k/mac.c  | 11 +++++++++++
 drivers/net/wireless/ath/ath12k/mhi.c  |  1 +
 drivers/net/wireless/ath/ath12k/peer.c |  6 ++++++
 drivers/net/wireless/ath/ath12k/qmi.c  |  1 +
 drivers/net/wireless/ath/ath12k/wmi.c  |  5 +++++
 8 files changed, 29 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index acf30b4..6355a92 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1225,6 +1225,8 @@ static int ath12k_core_start(struct ath12k_base *ab,
 		goto err_reo_cleanup;
 	}
 
+	WARN_ON(test_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags));
+	set_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags);
 	/* put hardware to DBS mode */
 	if (ab->hw_params->single_pdev_only) {
 		ret = ath12k_wmi_set_hw_mode(ab, WMI_HOST_HW_MODE_DBS);
@@ -1737,8 +1739,6 @@ static int ath12k_core_reconfigure_on_crash(struct ath12k_base *ab)
 	if (ret)
 		return ret;
 
-	clear_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
-
 	ret = ath12k_core_qmi_firmware_ready(ab);
 	if (ret)
 		goto err_hal_srng_deinit;
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 565af9c..5e3b8ac 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -239,6 +239,7 @@ enum ath12k_dev_flags {
 	ATH12K_FLAG_HW_GROUP_ATTACHED,
 	ATH12K_FLAG_FTM_SEGMENTED,
 	ATH12K_FLAG_PPE_DS_ENABLED,
+	ATH12K_FLAG_WMI_INIT_DONE,
 };
 
 enum ath12k_monitor_flags {
diff --git a/drivers/net/wireless/ath/ath12k/htc.c b/drivers/net/wireless/ath/ath12k/htc.c
index d8d0929..9e0aa73 100644
--- a/drivers/net/wireless/ath/ath12k/htc.c
+++ b/drivers/net/wireless/ath/ath12k/htc.c
@@ -643,6 +643,8 @@ int ath12k_htc_connect_service(struct ath12k_htc *htc,
 
 	reinit_completion(&htc->ctl_resp);
 
+	if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags))
+		clear_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
 	status = ath12k_htc_send(htc, ATH12K_HTC_EP_0, skb);
 	if (status) {
 		kfree_skb(skb);
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 4b10954..bfc501f 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -5691,6 +5691,16 @@ static int ath12k_mac_op_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 		return 0;
 	}
 
+	if (!(arvif && arvif->ar)) {
+		mutex_unlock(&ah->conf_mutex);
+		return -EINVAL;
+	}
+
+	if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &arvif->ar->ab->dev_flags)) {
+		mutex_unlock(&ah->conf_mutex);
+		return -ESHUTDOWN;
+	}
+
 	if (sta) {
 		ahsta = (struct ath12k_sta *)sta->drv_priv;
 		if (sta->mlo) {
@@ -7030,6 +7040,7 @@ static void ath12k_mac_ml_station_remove(struct ath12k_vif *ahvif,
 		ar = arvif->ar;
 
 		mutex_lock(&ar->conf_mutex);
+
 		ath12k_mac_station_post_remove(ar, arvif, arsta);
 		mutex_unlock(&ar->conf_mutex);
 
diff --git a/drivers/net/wireless/ath/ath12k/mhi.c b/drivers/net/wireless/ath/ath12k/mhi.c
index eacb7b5..8c4dd18 100644
--- a/drivers/net/wireless/ath/ath12k/mhi.c
+++ b/drivers/net/wireless/ath/ath12k/mhi.c
@@ -331,6 +331,7 @@ static void ath12k_mhi_op_status_cb(struct mhi_controller *mhi_cntrl,
 		break;
 	case MHI_CB_EE_RDDM:
 		if (!(test_bit(ATH12K_FLAG_UNREGISTERING, &ab->ag->dev_flags))) {
+			set_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
 			set_bit(ATH12K_FLAG_RECOVERY, &ab->dev_flags);
 			queue_work(ab->workqueue_aux, &ab->reset_work);
 		}
diff --git a/drivers/net/wireless/ath/ath12k/peer.c b/drivers/net/wireless/ath/ath12k/peer.c
index 126499a..0fdda6f 100644
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -389,6 +389,9 @@ int ath12k_ml_link_peers_delete(struct ath12k_vif *ahvif, struct ath12k_sta *ahs
 
 		cancel_work_sync(&arsta->update_wk);
 
+		if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags) ||
+				test_bit(ATH12K_FLAG_RECOVERY, &ar->ab->dev_flags))
+			continue;
 		mutex_lock(&ar->conf_mutex);
 		ath12k_dp_peer_cleanup(ar, arvif->vdev_id, arsta->addr);
 
@@ -414,6 +417,9 @@ int ath12k_ml_link_peers_delete(struct ath12k_vif *ahvif, struct ath12k_sta *ahs
 
 		ar = arvif->ar;
 
+		if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags) ||
+				test_bit(ATH12K_FLAG_RECOVERY, &ar->ab->dev_flags))
+			continue;
 		mutex_lock(&ar->conf_mutex);
 		ret = ath12k_wait_for_peer_delete_done(ar, arvif->vdev_id, arsta->addr);
 		if (ret) {
diff --git a/drivers/net/wireless/ath/ath12k/qmi.c b/drivers/net/wireless/ath/ath12k/qmi.c
index 4b0f576..2ee6404 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -5713,6 +5713,7 @@ static void ath12k_qmi_driver_event_work(struct work_struct *work)
 			break;
 		case ATH12K_QMI_EVENT_SERVER_EXIT:
 			set_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
+			clear_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags);
 			break;
 		case ATH12K_QMI_EVENT_REQUEST_MEM:
 			ret = ath12k_qmi_event_mem_request(qmi);
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index ca71791..3571bcd 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -540,6 +540,10 @@ int ath12k_wmi_cmd_send(struct ath12k_pdev_wmi *wmi, struct sk_buff *skb,
 	int ret = -EOPNOTSUPP;
 	struct ath12k_base *ab = wmi_sc->ab;
 
+	if (!(test_bit(ATH12K_FLAG_WMI_INIT_DONE, &wmi_sc->ab->dev_flags)) &&
+		cmd_id != WMI_INIT_CMDID)
+		return -ESHUTDOWN;
+
 	might_sleep();
 
 	if (ab->hw_params->credit_flow) {
@@ -13874,6 +13878,7 @@ void ath12k_wmi_detach(struct ath12k_base *ab)
 	for (i = 0; i < ab->htc.wmi_ep_count; i++)
 		ath12k_wmi_pdev_detach(ab, i);
 
+	clear_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags);
 	ath12k_wmi_free_dbring_caps(ab);
 }
 
-- 
2.7.4

